### ISSUE TITLE
Unit tests in main.rs (or submodule of main.rs) aren't detected

### ISSUE THREAD

**Issue by tinywombat765** [danielzfranklin/cargo-rr#5]
cargo rr test fails when running on a project using cargo workspaces. cargo-rr prouces the following error:
```
Error: cargo-rr: No test artifacts built.
```
I was testing on the following code, for reference: https://git.sr.ht/~zethra/license

**Comment by dzfranklin** [comment]
Thanks for the bug report. I've never used workspaces. I'll take a look, it's probably not too hard to fix

**Comment by dzfranklin** [comment]
The problem with zethra/license isn't that it's a workspace, it's that the unit test is compiled into a binary. The current design is based on my incorrect assumption that tests can either be defined in `/tests` or in `lib.rs` (or a module whose `mod` statement is in `lib.rs`).

It'll take me a little while to figure out a good fix for this. In the meantime you can install a version that disables "smartness" in guessing which targets can contain tests and just asks you to pick

```
cargo install --git https://github.com/danielzfranklin/cargo-rr.git --tag temp-disable-filters
```

**Comment by dzfranklin** [comment]
Tests in workspaces also seem to not be supported. As far as I can tell there are no tests in zethra/license not in the top-level crate, but when I insert one in one of the subcrates I get a different error. I've opened a separate issue for that in #6

**Comment by grunweg** [comment]
Just for the record: I just ran into this exact issue (unit test in main not detected).
Thanks for the workaround above, it seems to work for me. :heart:

**Comment by dzfranklin** [comment]
I want you to know I'm still working on this. I've figured out how to list tests across the whole project. Here's my current draft of the UX when you type `cargo rr test`

![Screenshot from 2021-05-18 14-55-16](https://user-images.githubusercontent.com/10361846/118664345-4c484600-b7e9-11eb-9cce-ec57a925db60.png)

**Comment by dzfranklin** [comment]
@grunweg @zethra I've refactored the project and written 41 test cases that I'm hoping cover all the different ways people write rust tests, examples, and binaries. If you're willing, I'd appreciate if you beta-tested the new version and let me know if it addresses your issues. You can install it with

```
cargo install --git https://github.com/danielzfranklin/cargo-rr.git --branch new-build-select
```

The one thing that isn't supported is `cargo rr run` with no arguments. That's waiting on [a cargo feature request](https://github.com/rust-lang/cargo/issues/9491).

Thanks!

**Comment by tinywombat765** [comment]
Sure, I'll give it a try

**Comment by grunweg** [comment]
@danielzfranklin Thanks for the prompt work on this! Sounds like you were in for quite a deep dive.

I just tried the new version. It's come a long way, but isn't fully there yet. More precisely:
- I had some failing tests I wanted to debug. Running `cargo t recognition` runs 10 tests, out of which two fail.
- Running `cargo rr test recognition` very quickly yields one line "Compiling...", lists the ten tests to run and then nothing happens. Am I supposed to do something to select a test? If so, what?
- The previous workaround worked for me, in that I got a run of failing tests which I could debug. What UX were you envisioning? I'm open for something else; I might be missing something.

**Comment by dzfranklin** [comment]
It's good to hear it's not clear @grunweg . The first test should be styled as selected. You should be able to use the up/down arrow keys to select a specific test, or type part of a test name to fuzzily filter the list. Then you hit enter and record only that test.

If this doesn't work please let me know. Otherwise I'd appreciate suggestions for making it clearer.

(This solves the fact that in a large project you often have multiple tests with similar names, and I only ever want to record one test. It's also faster to select a test by typing a few letters and hitting enter vs typing the unambiguous name and the selecting the module)

**Comment by grunweg** [comment]
Thanks for describing. That clears things up for me. I did in fact try this to see if it would work :-)
(In my setting, my tests are extremely short-running, so running only a specific test isn't a big issue for me. I understand that it would be in larger projects; in any case, being able to record only that test is a nice feature to have.) 

Sadly, the features doesn't work correctly for me yet: after selecting a test and hitting enter, I get an error message `invalid option --output-trace-dir`, followed by a display of `rr`'s CLI options and the message `cargo-rr: `rr record` exited with status exit status: 1. This whole output pops up almost immediately.

**Comment by grunweg** [comment]
Sorry for the long radio silence. I've found that my issue above was due to an old version of rr: at the time of my comment, I was running rr 5.2.0. Upgrading to version 5.4.0 (the version my distro is using) fixed it; now cargo-rr behaves as you described and works well enough for me.

@danielzfranklin Is there anything speaking against merging that branch into mainline?

**Comment by dzfranklin** [comment]
No problem @grunweg. I'm currently writing embedded stuff so I'm not using this. I just published `v0.2.0`.

**Comment by grunweg** [comment]
Wow, that's a fast response. (And you even published a new version already!) Thanks, @danielzfranklin!
